packages/pigeon/mock_handler_tester/test/message.dart

```diff
@@ -1,173 +1,148 @@
-    return result;
+  static SearchRequest decode(Object message) {
+    final Map<Object, Object> pigeonMap = message as Map<Object, Object>;
+    return SearchRequest()
+      ..query = pigeonMap['query'] as String
+      ..anInt = pigeonMap['anInt'] as int
+      ..aBool = pigeonMap['aBool'] as bool;
   }
 }
 
 class Nested {
   SearchRequest request;
+
   // ignore: unused_element
-  Map<dynamic, dynamic> _toMap() {
-    final Map<dynamic, dynamic> pigeonMap = <dynamic, dynamic>{};
-    pigeonMap['request'] = request == null ? null : request._toMap();
+  Object encode() {
+    final Map<Object, Object> pigeonMap = <Object, Object>{};
+    pigeonMap['request'] = request == null ? null : request.encode();
     return pigeonMap;
   }
 
   // ignore: unused_element
-  static Nested _fromMap(Map<dynamic, dynamic> pigeonMap) {
-    if (pigeonMap == null) {
-      return null;
-    }
-    final Nested result = Nested();
-    result.request = SearchRequest._fromMap(pigeonMap['request']);
-    return result;
+  static Nested decode(Object message) {
+    final Map<Object, Object> pigeonMap = message as Map<Object, Object>;
+    return Nested()
+      ..request = pigeonMap['request'] != null
+          ? SearchRequest.decode(pigeonMap['request'])
+          : null;
   }
 }
 
 abstract class FlutterSearchApi {
   SearchReply search(SearchRequest arg);
   static void setup(FlutterSearchApi api) {
     {
-      const BasicMessageChannel<dynamic> channel = BasicMessageChannel<dynamic>(
+      const BasicMessageChannel<Object> channel = BasicMessageChannel<Object>(
           'dev.flutter.pigeon.FlutterSearchApi.search', StandardMessageCodec());
-      channel.setMessageHandler((dynamic message) async {
-        final Map<dynamic, dynamic> mapMessage =
-            message as Map<dynamic, dynamic>;
-        final SearchRequest input = SearchRequest._fromMap(mapMessage);
-        final SearchReply output = api.search(input);
-        return output._toMap();
-      });
+      if (api == null) {
+        channel.setMessageHandler(null);
+      } else {
+        channel.setMessageHandler((Object message) async {
+          if (message == null) {
+            return null;
+          }
+          final SearchRequest input = SearchRequest.decode(message);
+          final SearchReply output = api.search(input);
+          return output.encode();
+        });
+      }
     }
   }
 }
 
 class NestedApi {
   Future<SearchReply> search(Nested arg) async {
-    final Map<dynamic, dynamic> requestMap = arg._toMap();
-    const BasicMessageChannel<dynamic> channel = BasicMessageChannel<dynamic>(
+    final Object encoded = arg.encode();
+    const BasicMessageChannel<Object> channel = BasicMessageChannel<Object>(
         'dev.flutter.pigeon.NestedApi.search', StandardMessageCodec());
-
-    final Map<dynamic, dynamic> replyMap = await channel.send(requestMap);
+    final Map<Object, Object> replyMap =
+        await channel.send(encoded) as Map<Object, Object>;
```

Comment:
Are these casts incurring any runtime cost?

Reply:
My understanding is that casts are cheaper than dynamic dispatch on `dynamic` these days, but I could be wrong.

Reply:
(To put it another way: there were casts before, this just made them visible.)

Reply:
(Specifically, before, we were assigning a `dynamic` (the return value of `await channel.send(encoded)`) to a variable of type `Map<dynamic, dynamic>`, which implies a cast. Now we are assigning an `Object` to a variable of type `Map<Object, Object>` with an explicit cast. I wouldn't be surprised if the assembly was actually byte for byte identical.